heroic-component/src/main/java/com/spotify/heroic/cluster/ClusterManager.java

```diff
@@ -63,9 +65,41 @@
      */
     AsyncFuture<Void> addStaticNode(URI node);
 
+    /**
+     * Eventually consistent view of the currently known nodes in the cluster
+     */
     List<ClusterNode> getNodes();
 
     /**
+     * Eventually consistent view of the currently known nodes in a specific shard
+     */
+    List<ClusterNode> getNodesForShard(Map<String, String> shard);
+
+    /**
+     * Run function on one random ClusterNode in a specific shard. The nodes referenced in the
+     * excludeIds list will not be considered.
+     *
+     * @param shard the shard
+     * @param excludeIds a list of node Ids that will not be considered when choosing a node
+     * @param fn the function to apply on the node
+     * @param <T> type of the return value
+     * @return Return value of the applied function, Id of the node, string representation of node
+     */
+    <T> Optional<NodeResult<T>> withNodeInShardButNotWithId(
+        Map<String, String> shard, List<UUID> excludeIds, Function<ClusterNode.Group, T> fn
+    );
```

Comment:
Unless we have a specific optimization for this we can take the exclusion criteria as a `Predicate<ClusterNode> exclude` to make it less specific?

The call would (according to my napkin coding) then be:
```java
withNodeInShardButNotWithId(shard, excludeIds::contains, fn)
```